pkg/cri/sbserver/container_create.go

```diff
@@ -279,6 +279,18 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
 		opts = append(opts, containerd.WithSandbox(sandboxID))
 	}
 
+	if c.nri.IsEnabled() {
```

Comment:
> Is the caller responsible for checking IsEnabled?

It should not be strictly necessary, as there is a test later in the call chain that checks and short-circuits the call at that level. Here I wanted to avoid scheduling the deferred function for possible failure cleanup and avoid adding the extra option which is unnecessary if NRI is disabled (and now that I look at it that option is ignorant about being disabled and relies on the other infra to properly not do anything then). 

> nri.API sometimes checks IsEnabled inside its associated functions, sometimes doesn't and we have nri.API's no-op implementation for non-Linux platforms.
> 
> Is it possible to use the no-op implementation for this disabled case?

Do you mean something like this ?
1. change `API` to be a interface not a struct
2. have a common no-op implementation of `API` (without any platform-specific build flags)
3. have platform-specific real implementations of `API`
4. if NRI is disabled instantiate the no-op implementation,
5. otherwise instantiate the real one
6. eliminate all/most of the server- an sbserver-level `if c.nri.IsEnabled()` checks

Or maybe instead of an interface, do the same thing with a 'frontend' struct that would then call to a real or a no-op implementation... but anyway, still mostly the same idea ?

If you think that would be a better approach (could be less intrusive/improve readability), I can give it a try and see what I manage to cook up...

Reply:
I wanted to remove IsEnabled checks, but this wouldn't cover `defer` case...

If NRI check is required for defer, how about using `nri != nil` instead of IsEnabled? Is it not coming from your change though.

Reply:
If we make `UndoCreateContainer()` to properly become a no-op when NRI is disabled, then it should be fine to unconditionally defer calling it on the error return path. There are a few places where currently we need to do an extra sandbox store lookup for the container's sandbox on the error branch if NRI is enabled. But maybe if we did those lookups from within the called NRI API function then we could make those calls unconditionally, too.

Anyway, I'll check if I can check if I could remove the top-level IsEnabled() checks in some safe way.

Reply:
@kzys I updated this PR, moving the enabled/disabled checks from `server` and `sbserver` to the common, split out bits. PTAL.